crates/ra_db/src/input.rs

```diff
@@ -111,11 +112,56 @@ impl CrateGraph {
     pub fn dependencies<'a>(
         &'a self,
         crate_id: CrateId,
-    ) -> impl Iterator<Item = &'a Dependency> + 'a {
+    ) -> impl Iterator<Item=&'a Dependency> + 'a {
         self.arena[&crate_id].dependencies.iter()
     }
+    fn dfs_find(&self, target: CrateId, from: CrateId) -> bool {
+        for dep in self.dependencies(from) {
+            let crate_id = dep.crate_id();
+            if crate_id == target {
+                return true;
+            }
+            if self.arena.contains_key(&crate_id) {
+                if self.dfs_find(target, crate_id) {
```

Comment:
This is not strictkly dfs, because we don't track visited items and hence can traverse the same node multiple times. 

If A depends on B and C, and both B and C depend on D, then we'll visit D's dependencies twice. This can lead to exponentialy bad performace, works case. Let's add a `visited: &mut FxHashSet<CrateId>` to prevent such situations?